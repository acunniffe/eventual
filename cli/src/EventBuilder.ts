import Ajv from 'ajv';
import pascalCase from 'pascal-case';
import * as path from 'path';

export function Event(
  rawSentence: string,
  lockKey: string,
  isDraft: boolean = false,
  filePath: string,
  line: number,
  overrideSchema?: any
): IEventBuilder {
  const tags = extractTags(rawSentence);
  const { variables, templateString } = replaceVariablesWithTemplateSyntax(
    rawSentence
  );

  const eventName = (() => {
    const mock: any = {};
    variables.map((i) => (mock[i] = i));

    const slug = pascalCase(rawSentence);
    return slug;
  })();

  const hasOverrideSchema = typeof overrideSchema === 'object';
  const schema = hasOverrideSchema
    ? overrideSchema
    : buildDefaultSchema(variables);

  const event: IEventBuilder = {
    eventName,
    lockKey,
    isDraft,
    tags,
    line,
    filePath,
    variables,
    rawSentence,
    toSentenceCode: buildSentenceCode(templateString, variables),
    schema,
  };

  return event;
}

// Regexes
const metaParserTemplateString = /([`"'])(?:(?=(\\?))\2.)*?\1/;
const tagRegex = /#(\w+)/g;
const simpleVariableRegex = /\$([\w]+)/g;

function buildSentenceCode(templateString: string, variables: string[]) {
  const arg = `{${variables.join(', ')}}`;
  return `(${arg}) => \`${templateString}\``;
}

export function buildDefaultSchema(variables: string[]) {
  const properties: { [key: string]: any } = {};
  variables.forEach((key) => (properties[key] = { type: 'string' }));
  return {
    type: 'object',
    required: variables,
    additionalProperties: false,
    properties,
  };
}

export function extractTags(rawSentence: string): string[] {
  const groups = [...rawSentence.matchAll(tagRegex)];
  return Array.from(new Set([...groups.map((i) => i[1].toLowerCase())])).sort();
}

export function replaceVariablesWithTemplateSyntax(
  rawSentence: string
): { variables: string[]; templateString: string } {
  useStringPolyfills();

  let result = rawSentence;
  let variables: string[] = [];
  //reverse it so inserts don't break indexes
  const groups = [...rawSentence.matchAll(simpleVariableRegex)].reverse();

  groups.forEach((match) => {
    const start: number = match.index!;
    const end = start + match[0].length;

    variables.push(match[1]);

    //@ts-ignore
    result = result.insert(end, '}');
    //@ts-ignore
    result = result.insert(start + 1, '{');
  });

  return {
    templateString: result,
    variables: Array.from(new Set([...variables])).sort(),
  };
}

export interface IEventBuilder {
  toSentenceCode: string;
  rawSentence: string;
  filePath: string;
  line: number;
  lockKey: string;
  eventName: string;
  variables: string[];
  isDraft: boolean;
  tags: string[];
  schema: any;
}

export interface IMetaSentenceParse {
  variables: string[];
  tags: string[];
  sentenceRaw: string;
}

// /// Generator
// export function generateCodeTo(outputPathProvided: string) {
//   const resolvedOutputPath = path.join(outputPathProvided);
//   const lockPath = path.join(resolvedOutputPath, '.lock.json');
//   console.log('lock at: ' + lockPath);
//   const outputPath = path.join(resolvedOutputPath);
//   const lockFile = getLockFile(lockPath);
//   if (lockFile) {
//     const updatedEventDefinitions = updateEventNamesFromLockFile(
//       __DefinedEvents,
//       lockFile
//     );
//     const newLockFile = checkLockFile(updatedEventDefinitions, lockFile);
//
//     saveLockFile(newLockFile, lockPath);
//   } else {
//     saveLockFile(generateLockFile(__DefinedEvents), lockPath);
//   }
// }

// utils
function useStringPolyfills() {
  //@ts-ignore
  String.prototype.insert = function (
    index: number,
    fullString: string
  ): string {
    if (index > 0) {
      return (
        this.substring(0, index) +
        fullString +
        this.substring(index, this.length)
      );
    }

    return fullString + this;
  };
}
