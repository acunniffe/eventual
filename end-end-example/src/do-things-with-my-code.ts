import { AUserWithEmailAndUserIdWasCreated } from './domain-events';

const a = AUserWithEmailAndUserIdWasCreated.new({
  email: 'aidan@Useoptic.com',
  userId: '123',
});
