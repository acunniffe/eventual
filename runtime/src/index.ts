import Ajv from 'ajv';
const ajv = new Ajv();
import colors from 'colors';

export default class EventualEventType<T> implements IEventualEventType {
  public readonly eventName: string;
  public readonly lockKey: string;
  public readonly tags: string[];

  private readonly schema: any;
  private readonly _toSentence: (props: any) => string;

  constructor(
    eventName: string,
    toSentence: (props: any) => string,
    schema: any,
    tags: string[],
    lockKey: string
  ) {
    this.eventName = eventName;
    this.tags = tags;
    this.lockKey = lockKey;
    this._toSentence = toSentence;
    this.schema = schema;
  }

  validate(props: any): boolean {
    const validate = ajv.compile(this.schema);
    return Boolean(validate(props));
  }

  new(props: any): IEventualEvent<T> {
    if (this.validate(props)) {
      return {
        type: this.eventName,
        data: props,
        toSentence: () =>
          this._toSentence(new Proxy(props, variableNameEchoProxy(props))),
        toJson: () => {
          return {
            type: this.eventName,
            data: props,
          };
        },
      };
    } else {
      throw new Error(`Invalid props for Event Type ${this.eventName}`);
    }
  }
}

const variableNameEchoProxy = (target: any) => ({
  get: function (target: any, prop: string) {
    return `\$${prop} ${colors.bold.underline(target[prop])}`;
  },
});

export interface IEventualEventType {
  eventName: string;
  tags: string[];
  lockKey: string;
}

export interface IEventualEvent<T> {
  type: string;
  data: T;
  toSentence: () => string;
  toJson(): any;
}
