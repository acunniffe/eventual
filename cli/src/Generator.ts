import { IEventBuilder } from './EventBuilder';
import { compile } from 'json-schema-to-typescript';
import prettier from 'prettier';
import * as path from 'path';
export async function Generate(events: IEventBuilder[]) {
  const allEventCodePromises = events.map((i) => {
    return GenerateRuntimeDeclarationForEvent(i);
  });

  const resolved = await Promise.all(allEventCodePromises);
  const unformattedCode = `
${"import EventualEventType from '@eventual/runtime';"}
${resolved.join('')}`;
  return prettier.format(unformattedCode, {
    semi: false,
    parser: 'typescript',
  });
}

export async function GenerateRuntimeDeclarationForEvent(
  event: IEventBuilder
): Promise<string> {
  const tsInterfaceName = `I${event.eventName}`;
  const tsInterface = await compile(event.schema, tsInterfaceName, {
    bannerComment: '',
  });

  process.cwd();
  const leadingComment = `// [${event.tags}] ${
    event.rawSentence
  }\n// source: (Line ${event.line}) ${path.relative(
    process.cwd(),
    event.filePath
  )}`;

  const runtimeClassInstantiation = `export const ${
    event.eventName
  } = new EventualEventType<${tsInterfaceName}>('${event.eventName}', ${
    event.toSentenceCode
  }, ${JSON.stringify(event.schema)}, ${JSON.stringify(event.tags)}, '${
    event.lockKey
  }')`;

  return `\n\n\n${leadingComment}\n${tsInterface}\n\n${runtimeClassInstantiation}\n\n//////////////////////////////////////////////////////////`;
}
