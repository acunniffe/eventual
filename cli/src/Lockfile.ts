import path from 'path';
import { IEventBuilder } from './EventBuilder';
import * as fs from 'fs-extra';
//@ts-ignore
import deepEqual from 'deep-equal';
//@ts-ignore
import stringify from 'json-stable-stringify';

export function getLockFile(lockFilePath: string): ILockfile | undefined {
  if (fs.existsSync(lockFilePath)) {
    const lockFileContents: ILockfile = JSON.parse(
      fs.readFileSync(lockFilePath).toString()
    );
    return lockFileContents;
  }
}

export function saveLockFile(lockFile: ILockfile, lockFilePath: string) {
  fs.ensureFileSync(lockFilePath);
  fs.writeFileSync(lockFilePath, stringify(lockFile, { space: 1 }));
}

export function updateEventNamesFromLockFile(
  events: IEventBuilder[],
  lockFile: ILockfile
): IEventBuilder[] {
  return events.map((event) => {
    if (lockFile[event.lockKey]) {
      return { ...event, eventName: lockFile[event.lockKey].eventName };
    } else {
      return event;
    }
  });
}

export function checkLockFile(
  events: IEventBuilder[],
  currentLockFile: ILockfile
) {
  const newLockFile = generateLockFile(events);
  compareLockFiles(currentLockFile, newLockFile);
  return newLockFile;
}

export function compareLockFiles(base: ILockfile, head: ILockfile) {
  const { added, removed, intersect } = compareLockFileKeys(base, head);
  assertNoNonDraftRemovals(removed, base);
  assertNoSchemaChanges(intersect, base, head);

  return true;
}

function assertNoNonDraftRemovals(
  removed: string[],
  lockFileContents: ILockfile
) {
  const result: ILockfileEntry[] = Object.entries(lockFileContents)
    .filter((i) => removed.includes(i[0]))
    .map((i) => i[1]);

  const removedProds = result.filter((i) => !i.isDraft);
  if (removedProds.length > 0) {
    throw new Error(
      'You can not remove published events...ever! ' + removedProds.join(', ')
    );
  }
}
function assertNoSchemaChanges(
  intersection: string[],
  baseLockfile: ILockfile,
  headLockfile: ILockfile
) {
  intersection.forEach((lockKey) => {
    const baseSchema = baseLockfile[lockKey].schema;
    const headSchema = headLockfile[lockKey].schema;

    if (!deepEqual(baseSchema, headSchema)) {
      throw new Error(`Schema for ${lockKey} has changed! This is not allowed`);
    }
  });
}

export function compareLockFileKeys(
  base: ILockfile,
  head: ILockfile
): { removed: string[]; intersect: string[]; added: string[] } {
  const baseLockKeys = [...Object.keys(base)];
  const headLockKeys = [...Object.keys(head)];

  const removed = baseLockKeys.filter((x: string) => !headLockKeys.includes(x));
  const intersect = baseLockKeys.filter((x: string) =>
    headLockKeys.includes(x)
  );
  const added = headLockKeys.filter((x: string) => !baseLockKeys.includes(x));

  return {
    added,
    removed,
    intersect,
  };
}

export function generateLockFile(events: IEventBuilder[]): ILockfile {
  const contents: ILockfile = {};
  events.forEach(
    (i) =>
      (contents[i.lockKey] = {
        eventName: i.eventName,
        schema: i.schema,
        isDraft: i.isDraft,
      })
  );

  return contents;
}

interface ILockfile {
  [key: string]: ILockfileEntry;
}
interface ILockfileEntry {
  isDraft: boolean;
  eventName: string;
  schema: any;
}
