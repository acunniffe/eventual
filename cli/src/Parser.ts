import { Event, IEventBuilder } from './EventBuilder';
//@ts-ignore
import jsonic from 'jsonic';
//@ts-ignore
import niceTry from 'nice-try';
// Regexes
const matchAll = require('string.prototype.matchall');
matchAll.shim();

const getLinesFromMarkdown = /^((?:DRAFT |))\[([A-Z]{2,}[-][0-9]{1,5})\](.*)(?:\n```[a-z]*\n([\s\S]*?)\n```|)/gm;
const simpleVariableRegex = /\$([\w]+)/g;

export function extractRawTagsFromMarkdown(
  filePath: string,
  fileContents: string
): ITagParseResult[] {
  const groups = [...fileContents.matchAll(getLinesFromMarkdown)];
  return groups.map((match) => {
    const isDraft = match[1].startsWith('DRAFT');
    const lockKey = match[2];
    const rawSentence = match[3].trim();

    const rawJson = match[4] ? match[4].trim() : undefined;
    const parsedJson = niceTry(() => jsonic(rawJson));

    console.log(rawJson);
    console.log(parsedJson);

    const line =
      Array.from(fileContents.substring(0, match.index!)).filter(
        (i) => i === '\n'
      ).length + 1;

    return {
      isDraft,
      filePath,
      lockKey,
      line,
      rawSentence,
      rawJson,
      parsedJson,
      eventBuilder: Event(
        rawSentence,
        lockKey,
        isDraft,
        filePath,
        line,
        parsedJson
      ),
    };
  });
}

export interface ITagParseResult {
  isDraft: boolean;
  filePath: string;
  line: number;
  lockKey: string;
  rawSentence: string;
  eventBuilder: IEventBuilder;
  rawJson: string | undefined;
  parsedJson: any | undefined;
}
