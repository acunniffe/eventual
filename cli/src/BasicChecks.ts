import { ITagParseResult } from './Parser';
import colors from 'colors';
//@ts-ignore
import groupBy from 'lodash.groupby';
export function checkDuplicateLockKeys(tags: ITagParseResult[]): string[] {
  const groups = groupBy(tags, 'lockKey');
  const duplicates = Object.entries(groups)
    .filter((i) => {
      const [key, events] = i;
      //@ts-ignore
      return events.length > 1;
    })
    .map((i) => i[0]);

  return duplicates;
}

export function checkCustomSchemas(tags: ITagParseResult[]): void {
  tags.forEach((i) => {
    if (i.rawJson && !i.parsedJson) {
      throw new Error(
        `Invalid JSON provided as custom schema to ${i.lockKey} at Line(${i.line}) ${i.filePath}`
      );
    }
    if (i.rawJson && i.parsedJson) {
      validateOverrideSchema(
        i.parsedJson,
        i.eventBuilder.variables,
        `${i.lockKey} at Line(${i.line}) ${i.filePath}`
      );
    }
  });
}

export function validateOverrideSchema(
  schema: any,
  variables: string[],
  errorLocation: string
) {
  if (schema.type === 'object') {
    const keys = Object.keys(schema.properties);
    const hasAllKeys = variables.every((i) => keys.includes(i));
    const hasAllRequired = variables.every((i) => schema.required.includes(i));
    if (!hasAllKeys) {
      throw new Error(
        `Schema must have properties for all variables in the sentence ${errorLocation} \n` +
          JSON.stringify(schema)
      );
    }
    if (!hasAllRequired) {
      throw new Error(
        `Schema must have required properties for all variables in the sentence ${errorLocation} \n` +
          JSON.stringify(schema)
      );
    }
  } else {
    throw new Error(
      `Schema must be an object ${errorLocation} \n` + JSON.stringify(schema)
    );
  }
}

export function printDuplicateKeys(
  tags: ITagParseResult[],
  duplicateKeys: string[]
) {
  duplicateKeys.forEach((key) => {
    console.log(
      colors.red(`Error: ${colors.underline(key)} used multiple times`)
    );
    tags
      .filter((i) => i.lockKey === key)
      .forEach((i) => {
        console.log(colors.red(`Line ${i.line} ${i.filePath}`));
      });

    console.log('\n\n');
  });
}
