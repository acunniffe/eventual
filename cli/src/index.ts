#!/usr/bin/env node
import { extractRawTagsFromMarkdown, ITagParseResult } from './Parser';

const { program } = require('commander');
const pJson = require('../package.json');
import fs from 'fs-extra';
import path from 'path';
//@ts-ignore
import flatten from 'lodash.flatten';
import colors from 'colors';
import {
  checkCustomSchemas,
  checkDuplicateLockKeys,
  printDuplicateKeys,
} from './BasicChecks';
import {
  checkLockFile,
  generateLockFile,
  getLockFile,
  saveLockFile,
  updateEventNamesFromLockFile,
} from './Lockfile';
import { Generate } from './Generator';
import { IEventBuilder } from './EventBuilder';
program.version(pJson.version);

//@ts-ignore
program
  .command('generate [files...]')
  .option(
    '-s, --stage',
    'pass this flag when you only want to verify, but not generate any code'
  )
  .option('-o, --output <file>', 'where the file should be generated')
  .action(async (files: string[], cmdObj: any) => {
    const outputFile = cmdObj.output || path.join(process.cwd(), 'output.ts');
    const outputDirectory = path.dirname(outputFile);
    //@ts-ignore
    return await processFiles(files, true, outputFile, outputDirectory);
  });

program.parse(process.argv);

async function processFiles(
  files: string[],
  generate: boolean,
  outputFile: string,
  outputDirectory: string
) {
  const allFiles: string[] = files.map((i: string) => {
    return path.resolve(path.join(process.cwd(), i));
  });

  const nonExistent: string[] = allFiles.filter(
    (i: string) => !fs.existsSync(i)
  );
  if (nonExistent.length > 0) {
    return console.log(
      colors.red.bold('File(s) not found ' + JSON.stringify(nonExistent))
    );
  }

  const allFilesContents = await Promise.all(
    allFiles.map((path) =>
      fs.readFile(path).then((contents) => [path, contents.toString()])
    )
  );

  const flattened = flatten(
    allFilesContents.map(([filePath, fileContents]) => {
      return extractRawTagsFromMarkdown(filePath, fileContents);
    })
  );

  //basic checks
  const duplicates = checkDuplicateLockKeys(flattened);
  if (duplicates.length > 0) {
    return printDuplicateKeys(flattened, duplicates);
  }

  try {
    checkCustomSchemas(flattened);
  } catch (e) {
    console.log(colors.red(e.message));
    console.log(
      colors.red.bold('\n\nTerminating. Please fix the schemas and try again')
    );
  }

  const flattenedEventBuilders = flattened.map(
    (i: ITagParseResult) => i.eventBuilder
  );

  //use lock file
  const lockPath = path.join(outputDirectory, 'eventual.lock.json');
  console.log(colors.bold('Looking for lock at: ' + lockPath));
  const lockFile = getLockFile(lockPath);
  if (lockFile) {
    console.log(colors.green('Lock file found at ' + lockPath));

    const updatedEventDefinitions = updateEventNamesFromLockFile(
      flattenedEventBuilders,
      lockFile
    );

    try {
      const newLockFile = checkLockFile(updatedEventDefinitions, lockFile);
      saveLockFile(newLockFile, lockPath);

      if (generate) {
        return generateCode(updatedEventDefinitions);
      }
    } catch (e) {
      console.log(colors.red(e.message));
      console.log(
        colors.red.bold('\n\nTerminating. You can not break your domain')
      );
    }
  } else {
    console.log(colors.yellow('No lock file found. Creating one now.'));
    saveLockFile(generateLockFile(flattenedEventBuilders), lockPath);
    if (generate) {
      return generateCode(flattenedEventBuilders);
    }
  }

  async function generateCode(updatedEventBuilders: IEventBuilder[]) {
    const code = await Generate(updatedEventBuilders);

    await fs.writeFile(outputFile, code);

    console.log(
      `\n\n${colors.bold.green('Success!\n')}Types for ${
        updatedEventBuilders.length
      } events generated to ${colors.bold(
        outputFile
      )}\nLock file updated ${colors.bold(lockPath)}`
    );
  }
}
