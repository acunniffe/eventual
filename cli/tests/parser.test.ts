import { extractRawTagsFromMarkdown } from '../src/Parser';
import { replaceVariablesWithTemplateSyntax } from '../src/EventBuilder';

const markdownExample = `
# Domain Events
These are the events in our domain

## User Management Events
Users in our domain can be created and deleted.

[USER-1] A #user with $email and $userId was created.
[USER-2] #user $userId verified their email
DRAFT [USER-3] #user $userId deleted their account

`;

const markdownExampleWithACustomSchemaThrownIn = `
# Domain Events
These are the events in our domain

## User Management Events
Users in our domain can be created and deleted.

[USER-1] A #user with $email and $userId was created.
[USER-2] #user $userId verified their email
\`\`\`json
{
  "type": "object",
  "properties": {
    "userId": {
      "type": "string"
    },
    "date": {
      "type": "string"
    }
  }
}
\`\`\`
DRAFT [USER-3] #user $userId deleted their account

`;

test('can extract eventual tags from markdown', () => {
  expect(
    extractRawTagsFromMarkdown('example-file.js', markdownExample)
  ).toMatchSnapshot();
});

test('can build executable toSentence from raw', () => {
  expect(
    replaceVariablesWithTemplateSyntax(
      'A #user with $email and $userId was created.'
    )
  ).toMatchSnapshot();
});

test('can extract eventual tags with custom schemas from markdown', () => {
  console.log(
    extractRawTagsFromMarkdown(
      'example-file.js',
      markdownExampleWithACustomSchemaThrownIn
    )
  );
});
