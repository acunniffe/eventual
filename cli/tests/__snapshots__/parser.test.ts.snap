// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can build executable toSentence from raw 1`] = `
Object {
  "templateString": "A #user with \${email} and \${userId} was created.",
  "variables": Array [
    "email",
    "userId",
  ],
}
`;

exports[`can extract eventual tags from markdown 1`] = `
Array [
  Object {
    "eventBuilder": Object {
      "eventName": "AUserWithEmailAndUserIdWasCreated",
      "filePath": "example-file.js",
      "isDraft": false,
      "line": 8,
      "lockKey": "USER-1",
      "rawSentence": "A #user with $email and $userId was created.",
      "schema": Object {
        "additionalProperties": false,
        "properties": Object {
          "email": Object {
            "type": "string",
          },
          "userId": Object {
            "type": "string",
          },
        },
        "required": Array [
          "email",
          "userId",
        ],
        "type": "object",
      },
      "tags": Array [
        "user",
      ],
      "toSentenceCode": "({email, userId}) => \`A #user with \${email} and \${userId} was created.\`",
    },
    "filePath": "example-file.js",
    "isDraft": false,
    "line": 8,
    "lockKey": "USER-1",
    "rawSentence": "A #user with $email and $userId was created.",
  },
  Object {
    "eventBuilder": Object {
      "eventName": "UserUserIdVerifiedTheirEmail",
      "filePath": "example-file.js",
      "isDraft": false,
      "line": 9,
      "lockKey": "USER-2",
      "rawSentence": "#user $userId verified their email",
      "schema": Object {
        "additionalProperties": false,
        "properties": Object {
          "userId": Object {
            "type": "string",
          },
        },
        "required": Array [
          "userId",
        ],
        "type": "object",
      },
      "tags": Array [
        "user",
      ],
      "toSentenceCode": "({userId}) => \`#user \${userId} verified their email\`",
    },
    "filePath": "example-file.js",
    "isDraft": false,
    "line": 9,
    "lockKey": "USER-2",
    "rawSentence": "#user $userId verified their email",
  },
  Object {
    "eventBuilder": Object {
      "eventName": "UserUserIdDeletedTheirAccount",
      "filePath": "example-file.js",
      "isDraft": true,
      "line": 10,
      "lockKey": "USER-3",
      "rawSentence": "#user $userId deleted their account",
      "schema": Object {
        "additionalProperties": false,
        "properties": Object {
          "userId": Object {
            "type": "string",
          },
        },
        "required": Array [
          "userId",
        ],
        "type": "object",
      },
      "tags": Array [
        "user",
      ],
      "toSentenceCode": "({userId}) => \`#user \${userId} deleted their account\`",
    },
    "filePath": "example-file.js",
    "isDraft": true,
    "line": 10,
    "lockKey": "USER-3",
    "rawSentence": "#user $userId deleted their account",
  },
]
`;
